File: internal/modules/gws/tasks.go
package gws

import "context"

type gTask struct {
	ID          string `json:"id"`
	Title       string `json:"title"`
	Notes       string `json:"notes"`
	Status      string `json:"status"` // "needsAction" | "completed"
	Due         string `json:"due"`
	Completed   string `json:"completed"` // RFC3339, present only on completed tasks
	WebViewLink string `json:"webViewLink"`
}

// TaskItem mirrors the TS TaskItem payload shape.
type TaskItem struct {
	ID          string `json:"id"`
	Title       string `json:"title"`
	Notes       string `json:"notes,omitempty"`
	Due         string `json:"due"`
	Completed   bool   `json:"completed"`
	CompletedAt string `json:"completedAt,omitempty"`
	URL         string `json:"url"`
}
type TasksData struct {
	Tasks []TaskItem `json:"tasks"`
}

type tasksConfig struct {
	Tasklist        string `json:"tasklist"`
	ShowCompleted   bool   `json:"showCompleted"`
	CompletedMaxAge string `json:"completedMaxAge"`
	Limit           int    `json:"limit"`
}

func normalizeTask(t gTask) TaskItem {
	title := t.Title
	if title == "" {
		title = "(no title)"
	}
	return TaskItem{
		ID: t.ID, Title: title, Notes: t.Notes, Due: t.Due,
		Completed: t.Status == "completed", CompletedAt: t.Completed, URL: t.WebViewLink,
	}
}

func fetchTasks(ctx context.Context, run jsonRunner, cfg tasksConfig) (TasksData, error) {
	var resp struct {
		Items []gTask `json:"items"`
	}
	if err := run(ctx, []string{
		"tasks", "tasks", "list",
		"--params", jsonArg(map[string]any{
			"tasklist":      cfg.Tasklist,
			"maxResults":    cfg.Limit,
			"showCompleted": cfg.ShowCompleted,
			"showHidden":    cfg.ShowCompleted, // completed tasks are hidden by default
		}),
	}, &resp); err != nil {
		return TasksData{}, err
	}
	// The API returns items in manual (`position`) order — preserve it.
	tasks := make([]TaskItem, 0, len(resp.Items))
	for _, t := range resp.Items {
		tasks = append(tasks, normalizeTask(t))
	}
	return TasksData{Tasks: tasks}, nil
}

// setTaskCompleted flips a task's completion via `gws tasks tasks patch`.
// Un-completing sends completed:null so the timestamp clears under patch
// semantics.
func setTaskCompleted(ctx context.Context, run jsonRunner, tasklist, taskID string, completed bool) error {
	var body map[string]any
	if completed {
		body = map[string]any{"status": "completed"}
	} else {
		body = map[string]any{"status": "needsAction", "completed": nil}
	}
	var out map[string]any
	return run(ctx, []string{
		"tasks", "tasks", "patch",
		"--params", jsonArg(map[string]any{"tasklist": tasklist, "task": taskID}),
		"--json", jsonArg(body),
	}, &out)
}

File: frontend/legacy-modules/gws/tasks.ts
import { gwsJson } from "./gws";
import type { TasksConfig, TasksData, TaskItem } from "./manifest";

type GTask = {
  id: string;
  title?: string;
  notes?: string;
  status?: string; // "needsAction" | "completed"
  due?: string;
  completed?: string; // RFC3339 timestamp, present only on completed tasks
  webViewLink?: string;
};
type TasksResp = { items?: GTask[] };

export function normalizeTask(t: GTask): TaskItem {
  return {
    id: t.id,
    title: t.title || "(no title)",
    notes: t.notes,
    due: t.due ?? "",
    completed: t.status === "completed",
    completedAt: t.completed ?? "",
    url: t.webViewLink ?? "",
  };
}

export async function fetchTasks(config: TasksConfig): Promise<TasksData> {
  const resp = await gwsJson<TasksResp>([
    "tasks", "tasks", "list",
    "--params", JSON.stringify({
      tasklist: config.tasklist,
      maxResults: config.limit,
      showCompleted: config.showCompleted,
      showHidden: config.showCompleted, // completed tasks are hidden by default
    }),
  ]);
  // The API returns items in manual (`position`) order — preserve it.
  const tasks = (resp.items ?? []).map(normalizeTask);
  return { tasks };
}

/**
 * Flip a task's completion via `gws tasks tasks patch`. Un-completing sends
 * `completed: null` so the timestamp clears under patch semantics.
 */
export async function setTaskCompleted(
  tasklist: string,
  taskId: string,
  completed: boolean,
): Promise<void> {
  const body = completed
    ? { status: "completed" }
    : { status: "needsAction", completed: null };
  await gwsJson<GTask>([
    "tasks", "tasks", "patch",
    "--params", JSON.stringify({ tasklist, task: taskId }),
    "--json", JSON.stringify(body),
  ]);
}
